src/Core/Domain/Product/Shop/Command/SetProductShopsCommand.php

```diff
@@ -47,23 +43,25 @@ class CopyProductToShopCommand
     private $sourceShopId;
 
     /**
-     * @var ShopId
+     * @var ShopId[]
      */
-    private $targetShopId;
+    private $shopIds;
 
     /**
      * @param int $productId
      * @param int $sourceShopId
-     * @param int $targetShopId
+     * @param int[] $shopIds
      */
     public function __construct(
         int $productId,
         int $sourceShopId,
-        int $targetShopId
+        array $shopIds
     ) {
         $this->productId = new ProductId($productId);
         $this->sourceShopId = new ShopId($sourceShopId);
-        $this->targetShopId = new ShopId($targetShopId);
+        $this->shopIds = array_map(static function (int $shopId): ShopId {
```

Comment:
In the constructor you can check that:
- shopIds is not empty
- sourceShopId is part of the shop IDs

It's better to detect invalid argument early, and those checks don't require DB checks so no need to handle them in the handler

Reply:
* empty check right :+1: 
* but source shop must be checked against initial shop ids taken from db, and the source shop will be skipped anyway if it exists in selectedShops array, so I don't see a point forcing to add the shop into "selectedShops" when its always filtered out in handler anyway.